### ISSUE TITLE
Feature request: Utility method for checking entity exists.

### ISSUE THREAD

**Issue by huysamen** [sebelga/gstore-node#127]
Like the `.get()` method, but maybe just select the key and return if it existed.  I will try and do the implementation sometime and make a PR.

**Comment by sebelga** [comment]
Hello,
Yes it could be useful but I don't think we can only select the key and I don't think it would make any performance difference. I guess it would be a wrapper around `.get()` that returns `true | false`, right?
Great if you could make a PR for this! :+1:

**Comment by huysamen** [comment]
No it makes a big difference to only select the key, rather than the entire object.  Only selecting the key counts as a small operation, which is always free, while selecting the entire entity counts towards your quota.  Also, for large entities, much less data.  I was thinking something along the lines of:

```
(await MyEntity.list({ select: "__key__", limit: 1 })).entities.length !== 0;
```

**Comment by sebelga** [comment]
I think we misunderstood each other :smile: In a query it _does_ make a difference. In a `get(Key)` it does not.  
I thought that what you wanted is to check if an entity (by Id) exists... It seems that what you want is to check if there is at least 1 entity of a certain Kind, am I right?

**Comment by huysamen** [comment]
Yeah, sorry, don't think I was clear.  Actually, I don't think that example was right, it should be more something like:

```
(await MyEntity
    .query()
    .select("__key__")
    .filter("__key__", "=", datastore.key("MyEntity", id))
    .run()
).entities.length !== 0;
```

so basically the most optimal way to check if a specific entity exists.

**Comment by sebelga** [comment]
I am probably missing something but why not just

```js
const doesExist = await MyEntity.get(id) !== null;
```

Of course this means having `Gstore({ errorOnEntityNotFound: false });`

We could have a wrapper like this (to not force the global setting)

```js
// model.js
exist(id, ancestors, namespace) {
    return this.get(id, ancestors, namespace)
        .then(entity => entity !== null)
        .catch((err) => {
            if (err.code === errorCodes.ERR_ENTITY_NOT_FOUND) {
                return false;
            }
            throw(err);
        });
}
```

**Comment by huysamen** [comment]
My thinking is that fetching just the key is a free datastore operation, while the above will count against your quota.

**Comment by sebelga** [comment]
Mmm not sure where you read this. I see here https://cloudplatform.googleblog.com/2016/03/Google-Cloud-Datastore-simplifies-pricing-cuts-cost-dramatically-for-most-use-cases.html

> Small ops (projections and keys-only queries) will stay the same in only charging a single read for the entire query.

**Comment by huysamen** [comment]
Ah ok, I think I might have misunderstood.  Could it not still be a good idea to do this with a key-only query though?  It is a lot more efficient than reading a potentially large entity (1MB vs a few bytes).

**Comment by sebelga** [comment]
The thing is, getting an entity by key is extremely fast and I am not sure a query can beat it. If you have entities of 1MB then it might be more efficient a query... but then it's business specific imho. It would be interesting to ask this in the google-cloud/datastore repo to have more insight.

**Comment by sebelga** [comment]
I was looking for something else and just bumped into it:
https://cloud.google.com/appengine/quotas

It seems that you were right, there is an unlimited amount of Small Operations for free in the Quota. So it might be interesting to investigate the performance difference bwt a key fetch and query (key only) to decide what's best. I guess it also depends on the traffic and total number of entities...

**Comment by micaww** [comment]
I wonder if a local Datastore instance would be adequate for a benchmark?
I'm not sure how accurately it emulates the real Cloud Datastore. Minus the
networking/replication overhead of course.

**Comment by sebelga** [comment]
Not sure, but seems to me that it would make more sense a benchmark against the _real_ Datastore :) I guess that we would need to test scenario with thousands of entities to see how the Query response time change.
